lib/internal/modules/cjs/loader.js

```diff
@@ -1337,8 +1340,15 @@ function wrapSafe(filename, content, cjsModuleInstance, codeCache) {
     return result;
   } catch (err) {
     if (process.mainModule === cjsModuleInstance) {
-      const { enrichCJSError } = require('internal/modules/esm/translators');
-      enrichCJSError(err, content, filename);
+      if (getOptionValue('--experimental-detect-module')) {
+        // For the main entry point, cache the source to potentially retry as ESM.
+        process.mainModule[kEntryPointSource] = content;
```

Comment:
Why does this live in `process.mainModule`, which can be modified? I think it should be fine to just put it in the `module.exports` of `cjs/loader`, or to make the side-effects clear, save it in a top-level variable and add a pair of accessors for it to the `module.exports`? Then in `run_main.js` you just get/modify the exports of `cjs/loader`.